### PR TITLE
[HIP] Refactor error handling in enqueue.cpp

### PR DESCRIPTION
**Pull request by aarongreig**
LLVM testing https://github.com/intel/llvm/pull/12481

### CODE REVIEW

**Review by hdelan — APPROVED** [review]
LGTM

**Review by jchlanda — APPROVED** [review]
Really like this, needless accumulation of error codes in `Result`-like variables, instead of straight out returns, and not using `UR_CHECK_ERROR` are one of my top pet hates. Thank you!

**Comment by aarongreig** [comment]
@npmiller please take a look when you can

**Review by npmiller — APPROVED** [review]
LGTM, sorry about the late review!

File: source/adapters/hip/enqueue.cpp [hunk 1]
```diff
@@ -1516,10 +1500,18 @@ urEnqueueUSMAdvise(ur_queue_handle_t hQueue, const void *pMem, size_t size,
...
         releaseEvent();
-        setErrorMessage("mem_advise is ignored as the advice argument is not "
-                        "supported by this device",
-                        UR_RESULT_SUCCESS);
-        return UR_RESULT_ERROR_ADAPTER_SPECIFIC;
+        // UR_RESULT_ERROR_INVALID_ENUMERATION is returned when using a valid
+        // but currently unmapped advice arguments as not supported by this
+        // platform. Therefore, warn the user instead of throwing and aborting
+        // the runtime.
+        if (Result == UR_RESULT_ERROR_INVALID_ENUMERATION) {
```
Comment (GeorgeWeb) [inline source/adapters/hip/enqueue.cpp:1507]: I may be missing something here but is this `if (Result == UR_RESULT_ERROR_INVALID_ENUMERATION)` meant to be checked again inside the above `if (Result == UR_RESULT_ERROR_INVALID_ENUMERATION)`? It is already true in this case. Is it meant to be wrapped in `if (Result != UR_RESULT_SUCCESS)` -> check if invalid else `throw`?

**Review by GeorgeWeb — COMMENTED** [review]
Generally looks good. Thank you for cleaning that up!
Just a little comment I was confused about.

File: source/adapters/hip/enqueue.cpp (hunk 1 above)
Comment (aarongreig) [reply source/adapters/hip/enqueue.cpp]: I think I messed up the rebase here, now it just falls through to `UR_CHECK_ERROR(Result)` if we aren't doing the adapter specific thing

Comment (GeorgeWeb) [reply source/adapters/hip/enqueue.cpp]: Yeah that makes sense to me now. Thanks, I actually thuoght it must've been a dodgy rebase by the looks of it.

**Review by GeorgeWeb — APPROVED** [review]

**Pull request merged**